### ISSUE TITLE
Hyperband Algorithm Example Reproduction

### ISSUE THREAD

**Issue by martsalz** [microsoft/nni#1567]
**Short summary about the issue/question**:

Currently I use the Hyperband algorithm. By the way, the example under `examples/trials/mnist-hyperband/` is not included in the master branch as mentioned in the documentation.

I summed up all n values from the table `(81+27+9+3+1+27+9+3+1+9+3+1+6+2+5)` and set them as `maxTrialNum = 187`. For `s = 4` NNI produces the same values as given in the table. From `s = 3` NNI behaves differently. For the 0th iteration, `n = 27` is specified in the table. Here, NNI creates a total of 34 instead of 27 trials. For the 1st iteration (`s = 3`) 11 instead of 9. For the 2nd and 3rd iterations the n values of NNI match the table again.

How can I reproduce the listed table?

**nni Environment**:
- nni version: v0.8-320-g421065b
- nni mode(local|pai|remote): local
- OS: CentOS Linux 7
- python version: Python 3.6
- is conda or virtualenv used?: virtualenv 
- is running in docker?: no

**Anything else we need to know**:

https://arxiv.org/pdf/1603.06560.pdf
https://github.com/microsoft/nni/blob/master/docs/en_US/Tuner/HyperbandAdvisor.md

![image](https://user-images.githubusercontent.com/12211578/65495028-4a17d200-deb6-11e9-9f14-093cb819f6af.png)
![image](https://user-images.githubusercontent.com/12211578/65495051-556afd80-deb6-11e9-83cd-4c10d96303d9.png)

```
authorName: default
experimentName: example_mnist
trialConcurrency: 1
maxExecDuration: 72h
maxTrialNum: 187
localConfig:
    useActiveGpu: true
#choice: local, remote, pai
trainingServicePlatform: local
searchSpacePath: search_space.json
#choice: true, false
useAnnotation: false
advisor:
  #choice: Hyperband
  builtinAdvisorName: Hyperband
  classArgs:
    #R: the maximum STEPS (could be the number of mini-batches or epochs) can be
    #   allocated to a trial. Each trial should use STEPS to control how long it runs.
    R: 81
    #eta: proportion of discarded trials
    eta: 3
    #choice: maximize, minimize
    optimize_mode: minimize
trial:
  command: python3 test.py
  codeDir: .
  gpuNum: 1
```

**Comment by QuanluZhang** [comment]
@martsalz thanks for your feedback. May I ask how did you calculate `n`? And did you use `trialConcurrency=1` in your experiment?

**Comment by martsalz** [comment]
Yes, `trialConcurrency=1 `

After the experiment was finished, I sorted the table by trial no.. Afterwards I checked the `TRIAL_BUDGET ` parameter for each single experiment. Then I counted how many trials have `TRIAL_BUDGET=81 `or how many have `TRIAL_BUDGET=27` etc..

**Comment by QuanluZhang** [comment]
@martsalz According to line 2 in Algorithm 1, `n_0` should be 34 when `s=3`, right? If the `s` in `(s+1)` (line 2) is `s_max`, then the `n_0` is 27.

**Comment by martsalz** [comment]
Yes, it is.

**Comment by martsalz** [comment]
But I still can't reproduce the table according to your instructions (replace in line 2 `s` in `(s+1)` with `s_max`). See:

i | s=4 | s=3  | s=2  | s=1  | s=0  |
---|---|---|---|---|---|
0 | 81, 1  | 27, 3  | 9, 9  | 3, 27  | 1, 81  |
1 | 27, 3  | 9, 9  | 3, 27  | 1, 81  |   |
2 | 9, 9  | 3, 27  | 1, 81  |   |   |
3 | 3, 27  | 1, 81  |   |   |   |
4 | 1, 81  |   |   |   |   |

The left value represents `n_i`, the right value `r_i.`

**Comment by PurityFan** [comment]
@martsalz thanks for your great finding!

After my investigation, the table is calculated by the following formula:
`n = math.ceil(B/(R*(s+1))) * (eta**s)`
which is different from the formula given in algorithm1:
`n = math.ceil(B/(R*(s+1)) * (eta**s))`
Due to the different order of math.ceil, they come to different result of `n = 27` and `n = 34`.

Algorithm 1 and the table are from the published paper (https://arxiv.org/pdf/1603.06560.pdf) of hyperband algorithm, so I suspect there is something wrong or misunderstanding in this paper.

In my opinion, the author of hyperband may prefer get the estimate after all the calculation (you can refer to the prove of Thm8 in reference), which is using `n = 34`. Hyperband in NNI also make this choice in our source code.

Let me know if you have any more questions.

**Comment by martsalz** [comment]
Thanks. What do you think the table above would look like? `(R=81, eta=3)`

Where did you find this line of code? I found the following under https://github.com/microsoft/nni/blob/master/src/sdk/pynni/nni/hyperband_advisor/hyperband_advisor.py (line 140):

`self.n = math.ceil((s_max + 1) * (eta**s) / (s + 1) - _epsilon) # pylint: disable=invalid-name`

**Comment by PurityFan** [comment]
Yes, line 140 you found is the place we calculate the value of `n` in NNI. You can calculate the new table by yourself using the pseudocode in Algorithm1, it may only have little difference :)

**Comment by martsalz** [comment]
Thanks. 

Then, the table should be changed in the documentation.
Back to my previous question/problem:

If you execute the Hyperband example with `R=81, eta=3, maxTrialNum=187` and `trialConcurrency=1`, are you able to reproduce the (new) table with NNI? - I can't.

> After the experiment was finished, I sorted the table by trial no.. Afterwards I checked the TRIAL_BUDGET parameter for each single experiment. Then I counted how many trials have TRIAL_BUDGET=81 or how many have TRIAL_BUDGET=27 etc..
> 
> I assumed that the trials ran from s=4 (starting with iteration 0 to iteration 4) to s=0. I.e. after (81+27+9+3+1) for s = 4 the next trial (the 122nd trial) belongs to s=3 and so on.

https://github.com/microsoft/nni/blob/master/docs/en_US/Tuner/HyperbandAdvisor.md


i | s=4 | s=3  | s=2  | s=1  | s=0  |
---|---|---|---|---|---|
0 | 81, 1  | 27, 3  | 9, 9  | **3**, 27  | **1**, 81  |
1 | 27, 3  | 9, 9  | 3, 27  | **1**, 81  |   |
2 | 9, 9  | 3, 27  | 1, 81  |   |   |
3 | 3, 27  | 1, 81  |   |   |   |
4 | 1, 81  |   |   |   |   |

**Comment by QuanluZhang** [comment]
@martsalz First, thanks for your suggestion, we will update the document to make it more clear.

Second, for the reproduction, I ran an experiment using your configuration. The table is as follows:
```
    4       3       2       1       0
    n   r   n   r   n   r   n   r   n   r
0   81  1   34  3   15  9   8   27  5   81
1   27  3   11  9   5   27  2   81
2   9   9   3   27  1   81
3   3   27  1   81
4   1   81
```
There are 206 trials in total. If you set `maxTrialNum=179`, you cannot get all the trials.
**Note that** the implementation of hyperband in NNI is optimized for high `trialConcurrency`. For example, there are only 3 trials when `s=3 and round=3`. If `trialConcurrency` is 10, there would be 7 resources left. So in the current implementation of hyperband in NNI, even though `s=3` is not finished, `s=2` will be started because `s=3` and `s=2` have not dependency. **Furthermore**, every time `s=[4,3,2,1,0]` is finished, we will start a new round of `s=[4,3,2,1,0]`. If `maxTrialNum` and `maxExecDuration` is large enough, `s=[4,3,2,1,0]` will be repeated infinitely. Maybe we should consider giving hyperband one more option, i.e., running `s=[4,3,2,1,0]` once or repeatedly.

**Comment by martsalz** [comment]
@QuanluZhang, thank you for the updated table.

Did you reproduce the table exactly with NNI (trial details)? Despite the fact that the hyperband algorithm in NNI is identical to the pseudocode in the paper, I cannot reproduce your table with NNI...

https://github.com/microsoft/nni/blob/master/src/sdk/pynni/nni/hyperband_advisor/hyperband_advisor.py (line 140)

**Comment by QuanluZhang** [comment]
@martsalz Here is how I generated the table: I added a log record in `get_n_r` function in hyperband_advisor.py as follow:
```diff
    def get_n_r(self):
        """return the values of n and r for the next round"""
        n = math.floor(self.n / self.eta**self.i + _epsilon)
        r = math.floor(self.r * self.eta**self.i + _epsilon)
+       _logger.debug('to debug: bracket %s next round %s, budget: %s', self.bracket_id, n, r)
        return n, r
```
This log record is written in `dispatcher.log`.

BTW, hyperband_advisor.py is the installed one, if you installed NNI locally, its location is `~/.local/lib/python3.x/site-packages/nni/hyperband_advisor/hyperband_advisor.py`.

**Comment by martsalz** [comment]
As I see it, the table is processed line by line with NNI, i.e. not each bracket separated but for each iteration for s=3...0 `n_i` and `r_i` are calculated. In the pseudocode each bracket is calculated for itself (iter. 0-x), afterwards the next bracket.

**Comment by martsalz** [comment]
Could you run the experiment completely and provide me the `experiment.json` file please?

**Comment by QuanluZhang** [comment]
@martsalz the experiment that I ran have 10000 trials, the file is too large. I also tried an experiment with 220 trials, and found that the out-of-order bracket&round (i.e., bracket 4 is not finished while bracket 3 is started even when trialConcurrrency is 1) is mainly due to the handling of multi-thread at the backend (i.e., msg_dispatcher_base). Thanks for your help, we will fix this bug soon.
